File: 144_Binary_Tree_Preorder_Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> r = new ArrayList<Integer>(); 
        return preOrderTraversal(root, r); 
    }
    
    public List<Integer> preOrderTraversal(TreeNode root, List<Integer> r)
    {
        if(root == null) return r; 
        r.add(root.val); 
        if(root.left != null) r = preOrderTraversal(root.left, r); 
        if(root.right != null) r = preOrderTraversal(root.right, r); 
        return r; 
    }
}

File: 144_Binary_Tree_Preorder_Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> r; 
        return preorderTraversal(root, r); 
    }
    
    vector<int> preorderTraversal(TreeNode* root, vector<int> r)
    {
        if(root == NULL) return r; 
        r.push_back(root->val); 
        if(root->left != NULL) r = preorderTraversal(root->left, r); 
        if(root->right != NULL) r = preorderTraversal(root->right, r); 
        return r;
    }
};